guides/source/active_record_callbacks.md

```diff
@@ -134,16 +173,197 @@ Here is a list with all the available Active Record callbacks, listed in the sam
+[`after_save`]:
+    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-after_save
+[`after_validation`]:
+    https://api.rubyonrails.org/classes/ActiveModel/Validations/Callbacks/ClassMethods.html#method-i-after_validation
+[`around_create`]:
+    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-around_create
+[`around_save`]:
+    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-around_save
+[`before_create`]:
+    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-before_create
+[`before_save`]:
+    https://api.rubyonrails.org/classes/ActiveRecord/Callbacks/ClassMethods.html#method-i-before_save
+[`before_validation`]:
+    https://api.rubyonrails.org/classes/ActiveModel/Validations/Callbacks/ClassMethods.html#method-i-before_validation
+
+There are examples below that show how to use these callbacks. We've grouped
+them by the operation they are associated with, and lastly showed how they can
+be used in combination. `after_commit` / `after_rollback` examples can be found
+[here](active_record_callbacks.html#after-commit-and-after-rollback).
+
+#### Validation Callbacks
+
+Validation callbacks are triggered whenever the record is validated directly via
+the
+[`valid?`](https://api.rubyonrails.org/classes/ActiveModel/Validations.html#method-i-valid-3F)
+or
+[`validate`](https://api.rubyonrails.org/classes/ActiveModel/Validations.html#method-i-valid-3F)
+methods, or indirectly via `create`, `update`, or `save`. They are called before
+and after the validation phase.
+
+```ruby
+class User < ApplicationRecord
+  validates :name, presence: true
+  before_validation :normalize_name
+  after_validation :check_errors
+
+  private
+    def normalize_name
+      self.name = name.downcase.titleize if name.present?
+      Rails.logger.info("Name normalized to #{name}")
+    end
+
+    def check_errors
+      if errors.any?
+        Rails.logger.error("Validation failed: #{errors.full_messages.join(', ')}")
+      end
+    end
+end
+```
+
+```irb
+irb> user = User.new(name: "", email: "john.doe@example.com", password: "abc123456")
+=> #<User id: nil, email: "john.doe@example.com", created_at: nil, updated_at: nil, name: "">
+
+irb> user.valid?
+Name normalized to
+Validation failed: Name can't be blank
+=> false
+```
+
+#### Save Callbacks
+
+Save callbacks are triggered whenever the record is persisted (i.e. "saved") to
+the underlying database, via the `create`, `update`, or `save` methods. They are
+called before, after, and around the object is saved.
+
+```ruby
+class User < ApplicationRecord
+  before_save :encrypt_password
+  around_save :log_saving
+  after_save :update_cache
+
+  private
+    def encrypt_password
+      self.password = BCrypt::Password.create(password)
+      Rails.logger.info("Password encrypted for user with email: #{email}")
```

Comment:
Some examples use `Rails.logger.info` and others use `puts` (in `log_destroy_action`).
I don't have a clear preference, `puts` is shorter, `Rails.logger.info` is what would be used in a real-world application, but we probably want to make these consistent?

Reply:
I read in a different Rails PR that it was advised to use `Rails.logger.info` over `puts`, hence I started using `Rails.logger.info`.  I spent some time looking for the referenced PR but can't find it 😞. I agree on the consistency, and maybe because it was mentioned to a contributor that `Rails.logger.info` is already the preferred way I'll keep using that and update the [Guides Guidelines updates - holding task](https://3.basecamp.com/3076981/buckets/35498807/todos/7116145100/edit?replace=true) task.